cloud/services/firewalls.go

```diff
@@ -0,0 +1,331 @@
+package services
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"slices"
+
+	"github.com/go-logr/logr"
+	"github.com/linode/linodego"
+
+	infrav1alpha1 "github.com/linode/cluster-api-provider-linode/api/v1alpha1"
+	"github.com/linode/cluster-api-provider-linode/util"
+)
+
+const (
+	maxFirewallRuleLabelLen = 32
+	maxIPsPerFirewallRule   = 255
+	maxRulesPerFirewall     = 25
+)
+
+var (
+	errTooManyIPs         = errors.New("too many IPs in this ACL, will exceed rules per firewall limit")
+	errDuplicateFirewalls = errors.New("duplicate firewalls found")
+)
+
+// HandleFirewall takes the CAPL firewall representation and uses it to either create or update the Cloud Firewall
+// via the given linode client
+func HandleFirewall(
+	ctx context.Context,
+	firewall *infrav1alpha1.LinodeFirewall,
+	linodeClient *linodego.Client,
+	logger logr.Logger,
+) (linodeFW *linodego.Firewall, err error) {
+	clusterUID := firewall.Spec.ClusterUID
+	linodeFWs, err := fetchFirewalls(ctx, firewall.Name, *linodeClient)
+	if err != nil {
+		logger.Info("Failed to list Firewalls", "error", err.Error())
+
+		return nil, err
+	}
+
+	// firewall conflict
+	if len(linodeFWs) > 1 {
+		logger.Info("Multiple firewalls found", "error", errDuplicateFirewalls.Error())
+
+		return nil, errDuplicateFirewalls
+	}
+
+	// build out the firewall rules for create or update
+	fwConfig, err := processACL(firewall, []string{clusterUID})
+	if err != nil {
+		logger.Info("Failed to process ACL", "error", err.Error())
+
+		return nil, err
+	}
+
+	if len(linodeFWs) == 0 {
+		logger.Info(fmt.Sprintf("Creating firewall %s", firewall.Name))
+		linodeFW, err = createFirewall(ctx, linodeClient, fwConfig)
+		if err != nil {
+			logger.Info("Failed to create firewall", "error", err.Error())
+
+			return nil, err
+		}
+		if linodeFW == nil {
+			err = errors.New("nil firewall")
+			logger.Error(err, "Created firewall is nil")
+
+			return nil, err
+		}
+	} else {
+		logger.Info(fmt.Sprintf("Updating firewall %s", firewall.Name))
+
+		linodeFW = &linodeFWs[0]
+		if err = updateFirewall(ctx, linodeClient, linodeFW, clusterUID, fwConfig); err != nil {
+			logger.Info("Failed to update firewall", "error", err.Error())
+
+			return nil, err
+		}
+	}
+
+	// Need to make sure the firewall is appropriately enabled or disabled after
+	// create or update and the tags are properly set
+	var status linodego.FirewallStatus
+	if firewall.Spec.Enabled {
+		status = linodego.FirewallEnabled
+	} else {
+		status = linodego.FirewallDisabled
+	}
+	if _, err = linodeClient.UpdateFirewall(
+		ctx,
+		linodeFW.ID,
+		linodego.FirewallUpdateOptions{
+			Status: status,
+			Tags:   util.Pointer([]string{clusterUID}),
```

Comment:
```suggestion
			Tags:   &[]string{clusterUID},
```
I think we only need `Pointer` functions for value types right?